lib/adapters/tool/playwright/index.ts

```diff
@@ -0,0 +1,311 @@
+                queue.onIdle().then(() => resolve(!code));
+            });
+        });
+    }
+
+    // Can't handle test results here because pwt does not provide api for this, so save instances and use them in custom reporter
+    handleTestResults(reportBuilder: GuiReportBuilder, eventSource: EventSource): void {
+        this._reportBuilder = reportBuilder;
+        this._eventSource = eventSource;
+    }
+
+    updateReference(): void {}
+
+    halt(err: Error): void {
+        logger.error(err);
+        process.exit(1);
+    }
+}
+
+function readPwtConfig(opts: ToolAdapterOptionsFromCli): {configPath: string, config: FullConfig} {
+    const configPaths = opts.configPath ? [opts.configPath] : DEFAULT_CONFIG_PATHS;
+    let originalConfig!: FullConfig;
+    let resolvedConfigPath!: string;
+
+    const revertTransformHook = setupTransformHook();
+
+    for (const configPath of configPaths) {
+        try {
+            resolvedConfigPath = path.resolve(configPath);
+            // eslint-disable-next-line @typescript-eslint/no-var-requires
+            const configModule = require(resolvedConfigPath);
+            originalConfig = configModule.__esModule ? configModule.default : configModule;
+
+            break;
+        } catch (err) {
+            if ((err as NodeJS.ErrnoException).code !== 'MODULE_NOT_FOUND') {
+                throw err;
+            }
+        }
+    }
+
+    revertTransformHook();
+
+    if (!originalConfig) {
+        throw new Error(`Unable to read config from paths: ${configPaths.join(', ')}`);
+    }
+
+    return {config: originalConfig, configPath: resolvedConfigPath};
+}
+
+async function registerTestResult(eventMsg: PwtEventMessage, reportBuilder: GuiReportBuilder): Promise<TestBranch> {
+    const {test, result, browserName, titlePath} = eventMsg;
+
+    const testCase = {
+        ...test,
+        titlePath: () => titlePath,
+        parent: {
+            ...test.parent,
+            project: () => ({
+                name: browserName
+            })
+        }
+    } as TestCase;
+
+    const testResult = {
+        ...result,
+        startTime: new Date(result.startTime)
+    } as TestResult;
+
+    const formattedResultWithoutAttempt = PlaywrightTestResultAdapter.create(testCase, testResult, UNKNOWN_ATTEMPT);
+    const formattedResult = await reportBuilder.addTestResult(formattedResultWithoutAttempt);
+
+    return reportBuilder.getTestBranch(formattedResult.id);
+}
+
+function prepareRunArgs(tests: TestSpec[], configPath: string, haveProjectsInConfig: boolean): string[] {
+    const testNames = new Set<string>();
+    const browserNames = new Set<string>();
+
+    for (const {testName, browserName} of tests) {
+        testNames.add(testName);
+        browserNames.add(browserName);
+    }
+
+    const args = ['test', '--reporter', path.resolve(__dirname, './reporter'), '--config', configPath];
+
+    if (testNames.size > 0) {
+        args.push('--grep', Array.from(testNames).join('|'));
```

Comment:
Will this work fine if test name contains brackets, etc or do we need additional escaping?